homeassistant/components/systemmonitor/coordinator.py

```diff
@@ -0,0 +1,148 @@
+"""DataUpdateCoordinators for the System monitor integration."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Iterator
+from datetime import datetime
+import logging
+import os
+from typing import TypeVar
+
+import psutil
+from psutil._common import sdiskusage, shwtemp, snetio, snicaddr, sswap
+from psutil._pslinux import svmem
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_component import DEFAULT_SCAN_INTERVAL
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+dataT = TypeVar(
+    "dataT",
+    bound=bool
+    | datetime
+    | dict[str, list[shwtemp]]
+    | dict[str, list[snicaddr]]
+    | dict[str, snetio]
+    | float
+    | int
+    | Iterator[psutil.Process]
+    | sswap
+    | str
+    | svmem
+    | tuple[float, float, float]
+    | sdiskusage
+    | None,
+)
+
+
+class MonitorCoordinator(DataUpdateCoordinator[dataT]):
```

Comment:
This seems like its going to create a lot more executor jobs which is one of the problems we had with the original design.

Maybe its better to have each sensor flip on parts of the update when its added to hass ?

Reply:
After the first refresh the coordinator won't update if there are no enabled entities.

Reply:
Oh right, it probably won't matter that much in that case.

Reply:
We already have protection built into the coordinators to avoid accidental synchronization so we shouldn't have a thundering herd problem either